[WIP] Python C-Reduce

mpflanzer opened the pull request.

regehr commented:
It looks like the CMake work needs to be done, will that be difficult?

regehr commented on creduce.py:
```diff
@@ -0,0 +1,167 @@
+
+if __name__ == "__main__":
+    try:
+        core_count = multiprocessing.cpu_count()
```
I can't find where core_count is actually used.

mpflanzer replied:
In line 59 `core_count` is used to set the default value for the `--n` argument

regehr commented:
The main executable should be called creduce, not creduce.py -- we don't want the users to care what language we're using.

regehr commented:
The python creduce doesn't run at all on Ubuntu 14.04 (which ships with Python 3.4.3) see the error below. Please make the python version requirements clear.

```
regehr@hawking:~/creduce-moritz$ ./creduce.py --help
Traceback (most recent call last):
  File "/home/regehr/creduce-moritz/creduce/creduce.py", line 40, in load_pass_group_file
    pass_group_dict = json.load(pass_group_file)
  File "/usr/lib/python3.4/json/__init__.py", line 268, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.4/json/__init__.py", line 318, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.4/json/decoder.py", line 343, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.4/json/decoder.py", line 359, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Expecting ',' delimiter: line 160 column 5 (char 7827)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./creduce.py", line 80, in <module>
    passes_group.add_argument("--pass-group", type=str, choices=get_available_pass_groups(), help="Set of passes used during the reduction")
  File "./creduce.py", line 41, in get_available_pass_groups
    pass_group_dict = CReduce.load_pass_group_file(path)
  File "/home/regehr/creduce-moritz/creduce/creduce.py", line 41, in load_pass_group_file
    except json.JSONDecodeError:
AttributeError: 'module' object has no attribute 'JSONDecodeError'
regehr@hawking:~/creduce-moritz$ python3 --version
Python 3.4.3
```

regehr commented on creduce/pass_groups/all.json:
```diff
@@ -0,0 +1,181 @@
+    {"pass": "ints", "arg": "c"},
+    {"pass": "ints", "arg": "d"},
+    {"pass": "ints", "arg": "e"},
+    {"pass": "balanced", "arg": "parens-only"}
```
missing comma

regehr commented on creduce/pass_groups/all.json:
```diff
@@ -0,0 +1,181 @@
+    {"pass": "clex", "arg": "rename-toks"},
+    {"pass": "clex", "arg": "delete-string"},
+    {"pass": "clex", "arg": "shorten-string"},
+    {"pass": "clex", "arg": "x-string"},
```
spurious comma

regehr commented:
I"d prefer that the output seen by users is same as the perl version, as opposed to this sort of thing:
```
@INFO@root@(78.9%, 1039033 bytes)
```

regehr commented on creduce/get_key.py:
```diff
@@ -0,0 +1,40 @@
+
+    i = 0
+
+    while True:
```
what's this busy loop for?

mpflanzer replied:
Opps, that file was not supposed to be committed. I used it to experiment with the "read key to abort pass" feature. If I remember correctly I have not found a solution for that yet. At least not one that works for Linux and Windows.

Is that feature an absolute must have?

regehr replied:
It's not an absolute must-have, but on the other hand I'd been planning on adding some more keyboard shortcuts such as those appearing in @vgvassilev's pull req that toggles the print-diffs mode.

regehr replied:
Maybe we can support these in UNIX but not Windows? I'd be OK with that.

regehr commented on creduce/passes/abstract.py:
```diff
@@ -0,0 +1,37 @@
+    class Result(enum.Enum):
+        ok = 0
+        stop = 1
+        error = 2
```
are python enums strongly typed? if not, it would make more sense to use larger numbers for these, that are unlikely to occur naturally

mpflanzer replied:
Yes, they are strongly typed.

regehr commented on creduce/utils/testing.py:
```diff
@@ -0,0 +1,847 @@
+            return False
+
+        # (b) there are not already to many variants (FIXME: can potentionally be removed later),
+        if len(self._environments) > 200:
```
what is going on here? where does 200 come from?

mpflanzer replied:
Currently it is just an arbitrary number. I ran into the problem of too many open files since also the failed variants are kept until a successful one is found and everything is killed.
I think this problem could also occur in the Perl version but I will have to check the details.

regehr replied:
Interesting, I've never seen it happen in the Perl version but I think you're right that it could happen.

regehr replied:
Please add a comment about open files.

mpflanzer commented:
Regarding CMake: Coming up with some solution to install C-Reduce via CMake will not be really difficult. It's more a question of finding a good solution.

With Python the normal way to install a "package" (everything Python related in the creduce subdir) would be to use the `setuptools` module from Python and to write a small `setup.py` for that. This would install the package into the system's or user's Python directory. By doing it this way there is no need to modify the `PYTHONPATH` manually. The `creduce(.py)` script would still be installed/copied via CMake.

But there is a "problem" with this approach: While CMake could also trigger the `setuptools` installation it should not do so because the user expects the CMake script to install all files under the `CMAKE_PREFIX_PATH`. The Python files would be installed independent of that variable into some other location.

We could split up the setup process into (1) invoke `setuptools` to install the `creduce` package and (2) invoke CMake to install everything else.

Alternatives would be:
- Call `setuptools` from within the CMake script and inform the user what will be installed where
- Don't use setuptools and copy everything with CMake (just as now with Perl). But the user will have to set `PYTHONPATH` manually
- The same as above but (I think) we could "hardcode" the install path into the Python files (using the CMake script). Then `PYTHONPATH` is not needed.

I don't really like the third solution because it makes C-Reduce less flexible. Currently the Python `creduce` package (everything except the top-level command line script) can be loaded and used from any Python script. This makes it easy from user to integrate C-Reduce into existing Python frameworks -- maybe to automate the reduction process of multiple files or to automatically analyse results etc.

I think I would vote for the approach where the user calls CMake and `setuptools` independently. But I make it up to you to decide how we want to handle the install process.

mpflanzer commented:
Regarding the `.py` extension: Ok, it shouldn't be a problem to remove the extension. But for Windows the extension is necessary to be able to run the script directly from the command line -- Windows determines the file type and thus the program to launch only based on the extension.

Example:
- `.\creduce.py --help` works from the command line
- `.\creduce --help` doesn't work
- `python .\creduce.py --help` works
- `python .\creduce --help` works

mpflanzer commented:
Regarding Python requirements: The current version requires at least Python 3.5. It might be possible to rewrite it to be compatible with Python 3.4 but it potentially complicates things. Especially the processing handling has changed quite a bit between Python 3.4 and 3.5.

Do you think it is a mayor problem to depend on Python 3.5?

mpflanzer commented:
Regarding output formatting: I just left the default string of the logging module for now. We can easily modify it to be (almost) the same as in the Perl version.

mpflanzer commented on creduce/utils/testing.py:
```diff
@@ -0,0 +1,847 @@
+    MAX_EXTRA_DIRS = 25000
+    GIVEUP_CONSTANT = 50000
+
+    def __init__(self, test_runner, pass_statistic, test_cases, parallel_tests, no_cache, silent_pass_bug, die_on_pass_bug, print_diff, max_improvement, no_give_up, also_interesting):
```
In line 139 in `creduce.py` (the top-level script) `args.n` is passed as argument to the constructor of the test manager.

regehr replied:
Thanks, I'm bad at Python.

eeide commented:
Regarding installation:

> The same as above but (I think) we could "hardcode" the install path into the Python files (using the CMake script). Then PYTHONPATH is not needed.

Unlike @mpflanzer, I am satisfied by an installation approach that treats C-Reduce as a black-box, standalone thing, not a Python module. We can "join the Python ecosystem" later if we want—but I do not consider this to be a goal of the current effort.

eeide commented:
> Do you think it is a mayor problem to depend on Python 3.5?

I agree with @regehr that C-Reduce should work "out of the box," insofar as possible, on Ubuntu 14.04, which is still in widespread use. Unless it is really infeasible, it should work with the Python that is available by default on that platform.

eeide commented:
> Regarding the `.py` extension [...]

You can tell that neither @regehr nor I are regular Windows users :-). I guess that our current Windows-platform users just rename the top-level script to `creduce.pl`? I had not realized this.

Perhaps I will lean on some people in my group who have spiff new Windows laptops to help me learn about this.

eeide commented:
> Regarding output formatting: I just left the default string of the logging module for now. We can easily modify it to be (almost) the same as in the Perl version.

This sounds great. Thank you for doing this!

eeide commented:
Re: Python version.

From a quick check, I guess that on OS X 10.11.6 (El Capitan), the system-installed version of Python is 2.7.10. Ancient! So certainly, on OS X, people will need to install a much newer version of Python (and C-Reduce's configure scripts will need to find it, etc.).

The goal is always to reduce friction where it is possible to do so with reasonable effort. @mpflanzer, I think you need to tell John and me how difficult it is for Python C-Reduce to work out of the box with Python 3.4 under Ubuntu 14. We _can_ require people to install Python 3.5—just like we require people to get an up-to-date LLVM—but avoiding such requirements where possible helps to make C-Reduce most usable, for as many people as possible.

mpflanzer commented:
Seems like I was overly pessimistic regarding Python 3.4. It pushed some small changes that add features which are missing in Python 3.4.
They could be removed again without much effort when Python 3.5 is widely available.

Python 2.7.10 has been released in May 2015. So it's not that ancient. ;-) The thing is that Python 2 and Python 3 are still being developed in parallel. Though for new projects Python 3 is the recommended version to use.
Anyway, on OS X it is fairly easy to install Python 3.5 via Homebrew (or some other package manager).

regehr commented:
regarding cmake, I'm not too particular about the details of the solution as long as we don't make C-Reduce harder to install than it is.

regehr commented:
we definitely want the binary to be creduce on any platform where this is possible, fine if it's different on Windows.

regehr commented:
If OSX doesn't come with a python3 at all, I think it's fine for us to require that users install it. If we can work with 3.4 on Ubuntu 14.04 until that platform disappears, that would be strongly preferable. Installing C-Reduce is already more of a chore than it should be!

regehr commented:
In summary, I still don't see any showstoppers here, but rather a few minor issues that need ironing out. I still plan to investigate the cases where creduce-perl does a better job than creduce-python.

Eric, have you done your own review of the code? what do you think? One possibility is we could merge this sooner rather than later and use the Python version as the basis for the paper we're working on (assuming the London people haven't committed to experiments using the Perl version).

mpflanzer commented:
I finally had time to pick up the work on the Python version again.

From the comments above it seems like only the output formatting still needs to be changed. That's fairly easy.

Regarding the name of the C-Reduce executable I run into a slight problem. My Python scripts are in an directory named `creduce` which means there cannot be a file named `creduce` as well (in the build directory). The folder name defines the name under which the Python modules can be loaded into other Python scripts, e.g. `import creduce`. Since the already existing packages (Ubuntu, Homebrew etc.) are also named `creduce` it would be good to keep this name for Python as well (I think).

My current solution is to keep the main executable as `creduce.py` in the build directory but install it as `creduce`. Not optimal but at least it works. But I am open for suggestions how to solve this problem.

I have already done some testing on Linux and have not run into problems. I will do some Windows testing as well, and maybe OS X. However, it would be great if you have some spare time to run creduce yourself. I might know some things which are not obvious for a normal user.

Well, and the autotools files need to be changed. I don't know much about autotools so I would appreciate if you could give it a try. The CMake files should be easy to understand. Otherwise I will see what I can do.

eeide commented:
I am the person who does the most hacking on the Autoconf-based build system I suppose. I'll assign this issue to myself to remind me to look at this.

XVilka commented:
Can be this revived?

regehr commented:
> Can be this revived?

The issue is that none of the core C-Reduce developers/maintainers loves Python, or is very good at it...

eeide commented:
I suspect that this won't be revived in the near term by @regehr or @chenyang78 or me.

XVilka commented:
I see. Maybe it makes sense to close it then? It has conflicts and very old anyway

mpflanzer commented:
> I suspect that this won't be revived in the near term by @regehr or @chenyang78 or me.

I would have some time to pick it up again. But as @regehr mentioned...

> The issue is that none of the core C-Reduce developers/maintainers loves Python, or is very good at it...

If @regehr, @eeide and @chenyang78 think it is worth too spend some more time with this I'm happy to help (or even do most/all of the work). However, if they prefer stick to the existing implementation -- which I can understand to some degree -- I rather invest the time into some of my other projects. 😉

Maybe the most interesting thing about this PR is that I tested different strategies on how to handle the merging of the parallel test. I can't remember all the detail right now but I should have notes somewhere which I might be able to dig out if there is any interest. At that time I prepared this spreadsheet: https://docs.google.com/spreadsheets/d/1Q2iDzYGuVCe13DPazcJr3Z0Jp0Dh-YJW8hAP54In-RI/edit?usp=sharing

Also, if I remember correctly, I found a few minor bugs in the pass implementations which I fixed in the Python code. Unfortunately, I never came around to address in the original Perl code. Maybe if I'm bored I'll compare the Python and Perl versions to try to find want I changed.

If (at the moment) there is no real interest in anything mentioned in this PR feel free to close it.

marxin commented:
> If (at the moment) there is no real interest in anything mentioned in this PR feel free to close it.

That's a pity that there's not much interest in the Python transition. Me as everyday user of the tool, I would be happy to work on the project once it's converted to Python.
I don't have any experience with Perl and I personally don't like the language much.

mpflanzer commented:
I think pretty much everything of value that I did here has been integrated into https://github.com/marxin/cvise/tree/master. I guess time to close this one here. :wink:

marxin commented:
Yep! Apprecite your help!

regehr commented:
yes, glad you found a better outlet for your code @mpflanzer!

@marxin have you done any comparisons between C-Reduce and C-Vise lately? are there any areas where C-Reduce is still better?

regehr commented:
also @marxin is there an explanation of C-Vise's parallelization strategy written down somewhere? @DRMacIver and I were recently discussing ways to get better parallel speedup. we're pretty sure we know how but I'm curious how this compares to what you did in C-Vise

marxin commented:
> @marxin have you done any comparisons between C-Reduce and C-Vise lately? are there any areas where C-Reduce is still better?

Hey! I don't know as I've left my job where I used C-Vise on a daily basis. The clang_delta is stared in between the projects and so I don't see any area where the current C-Reduce can be better :P

marxin commented:
For details, please observe the C-Vise repository ;)
